neo/Ledger/MemoryPool.cs

```diff
@@ -4,162 +4,358 @@
+                if (ret != 0) return ret;
+
+                return Transaction.Hash.CompareTo(tx.Hash);
+            }
+            
+            public int CompareTo(PoolItem otherItem)
+            {
+                if (otherItem == null) return 1;
+                return CompareTo(otherItem.Transaction, otherItem.FeePerByte);
             }
         }
 
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_fee = new ConcurrentDictionary<UInt256, PoolItem>();
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_free = new ConcurrentDictionary<UInt256, PoolItem>();
+        private static readonly double MaxSecondsToReverifyHighPrioTx = (double) Blockchain.SecondsPerBlock / 3;
+        private static readonly double MaxSecondsToReverifyLowPrioTx = (double) Blockchain.SecondsPerBlock / 5;
+        
+        // These two are not expected to be hit, they are just safegaurds. 
+        private static readonly double MaxSecondsToReverifyHighPrioTxPerIdle = (double) Blockchain.SecondsPerBlock / 15;
+        private static readonly double MaxSecondsToReverifyLowPrioTxPerIdle = (double) Blockchain.SecondsPerBlock / 30;
+        
+        /// <summary>
+        /// Store all verified unsorted transactions currently in the pool
+        /// </summary>
+        private readonly ConcurrentDictionary<UInt256, PoolItem> _unsortedTransactions = new ConcurrentDictionary<UInt256, PoolItem>();
+        /// <summary>
+        ///  Stores the verified low priority sorted transactions currently in the pool
+        /// </summary>
+        private readonly SortedSet<PoolItem> _sortedLowPrioTransactions = new SortedSet<PoolItem>();
+        /// <summary>
+        /// Stores the verified high priority sorted transactins currently in the pool
+        /// </summary>
+        private readonly SortedSet<PoolItem> _sortedHighPrioTransactions = new SortedSet<PoolItem>();
 
+        /// <summary>
+        /// Store the unverified transactions currently in the pool.
+        ///
+        /// Transactions in this data structure were valid in some prior block, but may no longer be valid.
+        /// The top ones that could make it into the next block get verified and moved into the verified data structures
+        /// (_unsortedTransactions, _sortedLowPrioTransactions, and _sortedHighPrioTransactions) after each block.
+        /// </summary>
+        private readonly ConcurrentDictionary<UInt256, PoolItem> _unverifiedTransactions = new ConcurrentDictionary<UInt256, PoolItem>();
+        private readonly SortedSet<PoolItem> _unverifiedSortedHighPriorityTransactions = new SortedSet<PoolItem>();
+        private readonly SortedSet<PoolItem> _unverifiedSortedLowPriorityTransactions = new SortedSet<PoolItem>();
+        
+        private int MaxHighPriorityTxsPerBlock => Settings.Default.MaxTransactionsPerBlock 
+            - Settings.Default.MaxFreeTransactionsPerBlock;
```

Comment:
I think that we can have `Settings.Default.MaxTransactionsPerBlock` transactions in a block if there are no free transactions.

Reply:
I could rename it since it isn't really the max, it's just how many it takes after a block is verified. More unverified transactions will now be pulled in anyway when processing the Idle message, and it is the plugin code that decides how many of the verified transactions to take currently. Also, at the time that the transactions are re-verified after a block is persisted it is unknown how many new high priority transactions will come in during the time until the next block.

Reply:
I addressed this comment and just use Settings.Default.MaxTransactionsPerBlock directly for high priority transactions.